lib/scholar/cluster/hierarchical.ex

```diff
@@ -0,0 +1,476 @@
+      Represented by the heights of each clade in a dendrogram plot.
+      Determined by both the `:dissimilarity` and `:linkage` options.
+
+    * `num_points` (`pos_integer/0`) -
+      Number of points in the dataset.
+      Must be $\\geq 3$.
+
+    * `sizes` (`Nx.Tensor` with shape `{n - 1}`) -
+      `sizes[i]` is the size of clade `i`.
+      If clade `k` was created by merging clades `i` and `j`, then
+      `sizes[k] == sizes[i] + sizes[j]`.
+
+  ## Examples
+
+      iex> data = Nx.tensor([[2], [7], [9], [0], [3]])
+      iex> Hierarchical.fit(data)
+      %Scholar.Cluster.Hierarchical{
+        clades: Nx.tensor([[0, 4], [1, 2], [3, 5], [6, 7]]),
+        dissimilarities: Nx.tensor([1.0, 2.0, 2.0, 4.0]),
+        num_points: 5,
+        sizes: Nx.tensor([2, 2, 3, 5])
+      }
+  """
+  deftransform fit(%Nx.Tensor{} = data, opts \\ []) do
+    opts = NimbleOptions.validate!(opts, @fit_opts_schema)
+    dissimilarity = opts[:dissimilarity]
+    linkage = opts[:linkage]
+
+    dissimilarity_fun =
+      case dissimilarity do
+        # :precomputed -> &Function.identity/1
+        :euclidean -> &pairwise_euclidean/1
+      end
+
+    update_fun =
+      case linkage do
+        :average -> &average/6
+        # :centroid -> &centroid/6
+        :complete -> &complete/6
+        # :median -> &median/6
+        :single -> &single/6
+        :ward -> &ward/6
+        :weighted -> &weighted/6
+      end
+
+    dendrogram_fun =
+      case linkage do
+        # TODO: :centroid, :median
+        l when l in [:average, :complete, :single, :ward, :weighted] ->
+          &parallel_nearest_neighbor/2
+      end
+
+    n =
+      case Nx.shape(data) do
+        {n, _num_features} ->
+          n
+
+        other ->
+          raise ArgumentError,
+                "Expected a rank 2 (`{num_obs, num_features}`) tensor, found shape: #{inspect(other)}."
+      end
+
+    if n < 3 do
+      raise ArgumentError, "Must have a minimum of 3 data points, found: #{n}."
+    end
+
+    pairwise = dissimilarity_fun.(data)
+
+    if Nx.shape(pairwise) != {n, n} do
+      raise ArgumentError, "Pairwise must be a symmetric matrix."
+    end
+
+    {clades, diss, sizes} = dendrogram_fun.(pairwise, update_fun)
+
+    %__MODULE__{
+      clades: clades,
+      dissimilarities: diss,
+      num_points: n,
+      sizes: sizes
+    }
+  end
+
+  # Clade functions
+
+  defnp parallel_nearest_neighbor(pairwise, update_fun) do
+    {n, _} = Nx.shape(pairwise)
+    pairwise = Nx.broadcast(:infinity, {n}) |> Nx.make_diagonal() |> Nx.add(pairwise)
+    clades = Nx.broadcast(-1, {n - 1, 2})
+    sizes = Nx.broadcast(1, {2 * n - 1})
+    pointers = Nx.broadcast(-1, {2 * n - 2})
+    diss = Nx.tensor(:infinity, type: Nx.type(pairwise)) |> Nx.broadcast({n - 1})
+
+    {clades, _count, _pointers, _pairwise, diss, sizes} =
+      while {clades, count = 0, pointers, pairwise, diss, sizes}, count < n - 1 do
```

Comment:
```suggestion
    {clades, _} =
      while {clades, {count = 0, pointers, pairwise, diss, sizes}}, count < n - 1 do
```

Reply:
For these suggestions, is the idea that the "non-returns" get stuffed into an additional tuple? That way the match can ignore them with a single `_`?

Reply:
Exactly :))

Reply:
Great pattern! I like it :) I did have to do quite a bit of "tuple matching" to get this working.